client/components/prompt.jsx

```diff
@@ -24,11 +32,43 @@ export default class Prompt extends React.Component {
   }
 
   handleKeyDown(e) {
-    if (e.key === this.state.chars[this.state.currentIndex]) {
-      this.setState({ currentIndex: this.state.currentIndex + 1 });
-      this.setState({ wrong: false });
-    } else if (e.key !== this.state.chars[this.state.currentIndex] && e.keyCode !== 16) {
-      this.setState({ wrong: true });
+    if (!this.state.startTime) {
+      this.setState({ startTime: new Date() });
+    }
+    if (this.state.currentIndex >= this.state.chars.length || this.props.time === 0) {
+      if (!this.state.endTime) {
+        this.setState({ endTime: new Date() });
+      }
+      this.setState({ testFinished: true });
+    }
+    if (!this.state.testFinished) {
+      if (e.key === this.state.chars[this.state.currentIndex]) {
+        this.setState({ currentIndex: this.state.currentIndex + 1 });
+        this.setState({ wrong: false });
+      } else if (e.key !== this.state.chars[this.state.currentIndex] && e.keyCode !== 16) {
+        this.setState({ wrong: true });
+      }
+    }
+  }
+
+  openModal() {
+    if (!this.state.testFinished) {
+      return 'hidden';
+    } else {
+      return '';
+    }
+  }
+
+  onResetClick() {
+    this.setState({ currentIndex: 0 });
+    this.setState({ testFinished: false });
```

Comment:
You can set more than one state property at once. There are a few places you can do this.

```js

this.setState({ currentIndex: 0, testFinished: false });
```